Consolidate -webkit-text-decoration and text-decoration shorthands

vitorroriz opened the pull request:
#### 4dc830ec3f354ee59ebb466ea26e46672f0fe490
<pre>
Consolidate -webkit-text-decoration and text-decoration shorthands
<a href="https://rdar.apple.com/156011594">rdar://156011594</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296097">https://bugs.webkit.org/show_bug.cgi?id=296097</a>

Reviewed by Ryosuke Niwa and Tim Nguyen.

Currently, text-decoration can only handle text-decoration-line
values, effectively making it a &quot;alias&quot; for it, while
-webkit-text-decoration can handle only 3 of text-decoration longhands.

We are consolidating behavior here, making -webkit-text-decoration an alias
for text-decoration. The property will be the shorthand for all 4 longhands:
- text-decoration-line
- text-decoration-thickness
- text-decoration-style
- text-decoration-color

We need to rebase a bunch of editing-related tests. Since text-decoration was
effectively working as an alias to text-decoration-line, this behavior was assumed
in EditingStyle. The changes here won&apos;t really change how things look from a user&apos;s
point of view, but in some cases we will serialize, for example, inline style,
&quot;text-decoration: underline&quot; as &quot;text-decoration-line: underline&quot;, which actually
make us more aligned with other engine implementations.

While implementing this, I&apos;ve realized there is a bug when propagating isCurrentColor
for the animation style. This also affect other shorthands, like text-emphasis.
So we are rebasing the animation tests here and aiming to intestigate it at:
<a href="https://bugs.webkit.org/show_bug.cgi?id=296914">https://bugs.webkit.org/show_bug.cgi?id=296914</a>

This is based on the previous attempts from:

* Oriol Brufau (@Loirooriol):
-  <a href="https://commits.webkit.org/248000@main">https://commits.webkit.org/248000@main</a> (reverted)
* Tim Nguyen (@nt1m):
- <a href="https://github.com/WebKit/WebKit/pull/6466">https://github.com/WebKit/WebKit/pull/6466</a>
* Matthieu Dubet (@mdubet)
- <a href="https://github.com/WebKit/WebKit/pull/23121">https://github.com/WebKit/WebKit/pull/23121</a>

* LayoutTests/TestExpectations:
* LayoutTests/editing/deleting/delete-line-break-before-underlined-content-expected.txt:
* LayoutTests/editing/deleting/delete-line-break-before-underlined-content-live-range-expected.txt:
* LayoutTests/editing/deleting/deleting-line-break-preserves-underline-color-expected.txt:
* LayoutTests/editing/deleting/deleting-line-break-preserves-underline-color-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-div-from-span-with-style-expected.txt:
* LayoutTests/editing/deleting/merge-div-from-span-with-style-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-multiple-text-decoration-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-multiple-text-decoration-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-style-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-style-live-range-expected.txt:
* LayoutTests/editing/execCommand/inline-style-after-indentoutdent-expected.txt:
* LayoutTests/editing/execCommand/inline-style-after-indentoutdent.html:
* LayoutTests/editing/execCommand/toggle-mixed-text-decorations-expected.txt:
* LayoutTests/editing/execCommand/toggle-mixed-text-decorations.html:
* LayoutTests/editing/pasteboard/insert-text-decoration-expected.txt:
* LayoutTests/editing/pasteboard/insert-u-with-text-decoration-none-expected.txt:
* LayoutTests/editing/pasteboard/preserve-underline-color-expected.txt:
* LayoutTests/editing/style/push-down-implicit-styles-mac-expected.txt:
* LayoutTests/editing/style/push-down-implicit-styles-mac.html:
* LayoutTests/editing/style/push-down-inline-styles-expected.txt:
* LayoutTests/editing/style/push-down-inline-styles.html:
* LayoutTests/fast/css/style-enumerate-properties-expected.txt:
* LayoutTests/fast/css/style-enumerate-properties.html:
* LayoutTests/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-letter-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-line-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/delete_3001-4000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/forwarddelete_3001-4000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/multitest_8001-9000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_1-1000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_1001-2000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_1-1000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_1001-2000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
* LayoutTests/platform/glib/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::canUseShorthandForLonghand):
* Source/WebCore/css/html.css:
(abbr[title], acronym[title]):
* Source/WebCore/css/parser/CSSPropertyParserCustom.h:
(WebCore::CSS::PropertyParserCustom::consumeSingleShorthand): Deleted.
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::styleWithResolvedTextDecorations const):
(WebCore::EditingStyle::collapseTextDecorationProperties):
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::EditingStyle::mergeStyle):
(WebCore::EditingStyle::removeStyleFromRulesAndContext):
(WebCore::EditingStyle::removeEquivalentProperties):
(WebCore::reconcileTextDecorationProperties):
(WebCore::StyleChange::StyleChange):
(WebCore::setTextDecorationProperty):
(WebCore::StyleChange::extractTextStyles):
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializePreservingVisualAppearanceInternal):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::isDefaultOrIgnorableTextDecorationLine):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthand):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthandSerialization):

Canonical link: <a href="https://commits.webkit.org/298456@main">https://commits.webkit.org/298456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d9273e21aef86fa5efd9b92a1d4326fb59d0d87

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25728 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121603 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9526ede0-d3ed-4e1b-a861-9e62d48de805) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117431 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43790 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/66083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8cc70c50-5c5f-4193-88f9-b616b9ec9049) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118490 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28626 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103702 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68177 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27772 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21818 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65276 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98007 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124761 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42451 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31815 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99892 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/96334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24511 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41571 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42337 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47900 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/41827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45157 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/43543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserCustom.h:
```diff
@@ -272,19 +272,6 @@ inline bool PropertyParserCustom::consumeStandardSpaceSeparatedShorthand(CSSPars
     return true;
 }
 
-inline bool PropertyParserCustom::consumeSingleShorthand(CSSParserTokenRange& range, PropertyParserState& state, const StylePropertyShorthand& shorthand, PropertyParserResult& result)
```
You can remove this as well: https://searchfox.org/wubkat/rev/2326b7031d277e0dc380c41edb926888df928161/Source/WebCore/css/parser/CSSPropertyParserCustom.h#126

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -40,6 +40,8 @@
 #include "StylePropertyShorthand.h"
 #include "StylePropertyShorthandFunctions.h"
+#include "rendering/style/RenderStyle.h"
+#include "rendering/style/RenderStyleConstants.h"
```
nit: you can drop rendering/style

weinig replied:
Neither of these includes should be necessary. This file already utilizes RenderStyle directly quite a bit.

vitorroriz replied:
You are right, some annoying tool added it and I haven't realized. I have to turn that off, thanks.

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -10785,14 +10770,19 @@
                     "text-decoration-style",
-                    "text-decoration-color"
+                    "text-decoration-color",
+                    "text-decoration-thickness"
```
The order of these matter, and in the spec, it is listed as:

`<'text-decoration-line'> || <'text-decoration-thickness'> || <'text-decoration-style'> || <'text-decoration-color'>`

Please also remember to update the `parser-grammar-unused` field as well as the specification URL to point to https://www.w3.org/TR/css-text-decor-4/#propdef-text-decoration.

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -652,11 +659,17 @@ Ref<MutableStyleProperties> EditingStyle::styleWithResolvedTextDecorations() con
-        style->setProperty(CSSPropertyTextDecorationLine, CSSPrimitiveValue::create(CSSValueNone));
+    if (valueList.isEmpty()) {
+        style->removeProperty(CSSPropertyTextDecorationLine);
+        // style->setProperty(CSSPropertyTextDecorationLine, CSSPrimitiveValue::create(CSSValueNone));
```
left over comment

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1569,9 +1595,39 @@ template<typename T>
+            continue;
+
+        // Only the text-decoration longhands support serializing with the shorthand in all editing properties.
+        CSSPropertyID shorthandID = property.shorthandID();
```
```suggestion
        auto shorthandID = property.shorthandID();
```

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1569,9 +1595,39 @@ template<typename T>
+    HashSet<CSSPropertyID> alreadyHandled;
     for (auto property : *m_mutableStyle) {
-        if (style.propertyMatches(property.id(), property.value()))
+        CSSPropertyID id = property.id();
```
```suggestion
        auto id = property.id();
```

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1569,9 +1595,39 @@ template<typename T>
+        }
+        auto shorthand = shorthandForProperty(shorthandID);
+        bool canRemoveAllLonghands = true;
+        for (CSSPropertyID longhandID : shorthand) {
```
```suggestion
        for (auto longhandID : shorthand) {
```

nt1m commented on Source/WebCore/editing/EditingStyle.cpp:
```diff
@@ -1569,9 +1595,39 @@ template<typename T>
+                break;
+            }
+        }
+        for (CSSPropertyID longhandID : shorthand) {
```
```suggestion
        for (auto longhandID : shorthand) {
```

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,69 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+    bool hasDefaultTextDecorationLine = textDecorationLine.isEmpty();
+    bool hasDefaultTextDecorationThickness = textDecorationThickness == RenderStyle::initialTextDecorationThickness();
+    bool hasDefaultTextDecorationStyle = textDecorationStyle == RenderStyle::initialTextDecorationStyle();
+    bool hasDefaultTextDecorationColor = equalIgnoringSemanticColor(state.style.visitedDependentColor(CSSPropertyTextDecorationColor), state.style.color());
```
Shouldn't this compare unvisited colors on both sides? Can we make a test to make sure this isn't a privacy leak?

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,69 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+    auto textDecorationLine = state.style.textDecorationLine();
+    auto textDecorationThickness = state.style.textDecorationThickness();
+    auto textDecorationStyle = state.style.textDecorationStyle();
+    auto textDecorationColor = state.style.textDecorationColor();
```
nit: I don't think these variables help readability, we should just inline them in the 2 places they are used

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,69 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+    return CSSValueList::createCommaSeparated(WTFMove(list));
+}
+
+inline void ExtractorCustom::extractTextDecorationShorthandSerialization(ExtractorState& state, StringBuilder& builder, const CSS::SerializationContext& context)
```
same comments for this method

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,60 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+        ExtractorSerializer::serializeTextDecorationLine(state, builder, context, state.style.textDecorationLine());
+    if (!hasDefaultTextDecorationThickness) {
+        if (!builder.isEmpty())
+            builder.append(" "_s);
```
```suggestion
            builder.append(' ');
```

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,60 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+    }
+    if (!hasDefaultTextDecorationStyle) {
+        if (!builder.isEmpty())
+            builder.append(" "_s);
```
```suggestion
            builder.append(' ');
```

nt1m commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -2758,6 +2761,60 @@ inline void ExtractorCustom::extractTextBoxShorthandSerialization(ExtractorState
+    }
+    if (!hasDefaultTextDecorationColor) {
+        if (!builder.isEmpty())
+            builder.append(" "_s);
```
```suggestion
            builder.append(' ');
```

nt1m reviewed (approved):
In the various places you call `cssText`, `CSS::defaultSerializationContext()` was replaced with `{ }`. Are they both the same?

vitorroriz committed:
Consolidate -webkit-text-decoration and text-decoration shorthands
rdar://156011594
https://bugs.webkit.org/show_bug.cgi?id=296097

Reviewed by Ryosuke Niwa and Tim Nguyen.

Currently, text-decoration can only handle text-decoration-line
values, effectively making it a "alias" for it, while
-webkit-text-decoration can handle only 3 of text-decoration longhands.

We are consolidating behavior here, making -webkit-text-decoration an alias
for text-decoration. The property will be the shorthand for all 4 longhands:
- text-decoration-line
- text-decoration-thickness
- text-decoration-style
- text-decoration-color

We need to rebase a bunch of editing-related tests. Since text-decoration was
effectively working as an alias to text-decoration-line, this behavior was assumed
in EditingStyle. The changes here won't really change how things look from a user's
point of view, but in some cases we will serialize, for example, inline style,
"text-decoration: underline" as "text-decoration-line: underline", which actually
make us more aligned with other engine implementations.

While implementing this, I've realized there is a bug when propagating isCurrentColor
for the animation style. This also affect other shorthands, like text-emphasis.
So we are rebasing the animation tests here and aiming to intestigate it at:
https://bugs.webkit.org/show_bug.cgi?id=296914

This is based on the previous attempts from:

* Oriol Brufau (@Loirooriol):
-  https://commits.webkit.org/248000@main (reverted)
* Tim Nguyen (@nt1m):
- https://github.com/WebKit/WebKit/pull/6466
* Matthieu Dubet (@mdubet)
- https://github.com/WebKit/WebKit/pull/23121

* LayoutTests/TestExpectations:
* LayoutTests/editing/deleting/delete-line-break-before-underlined-content-expected.txt:
* LayoutTests/editing/deleting/delete-line-break-before-underlined-content-live-range-expected.txt:
* LayoutTests/editing/deleting/deleting-line-break-preserves-underline-color-expected.txt:
* LayoutTests/editing/deleting/deleting-line-break-preserves-underline-color-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-div-from-span-with-style-expected.txt:
* LayoutTests/editing/deleting/merge-div-from-span-with-style-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-multiple-text-decoration-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-multiple-text-decoration-live-range-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-style-expected.txt:
* LayoutTests/editing/deleting/merge-paragraph-from-span-with-style-live-range-expected.txt:
* LayoutTests/editing/execCommand/inline-style-after-indentoutdent-expected.txt:
* LayoutTests/editing/execCommand/inline-style-after-indentoutdent.html:
* LayoutTests/editing/execCommand/toggle-mixed-text-decorations-expected.txt:
* LayoutTests/editing/execCommand/toggle-mixed-text-decorations.html:
* LayoutTests/editing/pasteboard/insert-text-decoration-expected.txt:
* LayoutTests/editing/pasteboard/insert-u-with-text-decoration-none-expected.txt:
* LayoutTests/editing/pasteboard/preserve-underline-color-expected.txt:
* LayoutTests/editing/style/push-down-implicit-styles-mac-expected.txt:
* LayoutTests/editing/style/push-down-implicit-styles-mac.html:
* LayoutTests/editing/style/push-down-inline-styles-expected.txt:
* LayoutTests/editing/style/push-down-inline-styles.html:
* LayoutTests/fast/css/style-enumerate-properties-expected.txt:
* LayoutTests/fast/css/style-enumerate-properties.html:
* LayoutTests/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-letter-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-pseudo/first-line-allowed-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-decoration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/delete_3001-4000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/forwarddelete_3001-4000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/multitest_8001-9000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_1-1000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_1001-2000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_1-1000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_1001-2000-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
* LayoutTests/platform/glib/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::canUseShorthandForLonghand):
* Source/WebCore/css/html.css:
(abbr[title], acronym[title]):
* Source/WebCore/css/parser/CSSPropertyParserCustom.h:
(WebCore::CSS::PropertyParserCustom::consumeSingleShorthand): Deleted.
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::init):
(WebCore::EditingStyle::styleWithResolvedTextDecorations const):
(WebCore::EditingStyle::collapseTextDecorationProperties):
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::EditingStyle::mergeStyle):
(WebCore::EditingStyle::removeStyleFromRulesAndContext):
(WebCore::EditingStyle::removeEquivalentProperties):
(WebCore::reconcileTextDecorationProperties):
(WebCore::StyleChange::StyleChange):
(WebCore::setTextDecorationProperty):
(WebCore::StyleChange::extractTextStyles):
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(WebCore::ReplaceSelectionCommand::removeRedundantStylesAndKeepStyleSpanInline):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializePreservingVisualAppearanceInternal):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::isDefaultOrIgnorableTextDecorationLine):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthand):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthandSerialization):

Canonical link: https://commits.webkit.org/298456@main
